File: android/app/src/main/kotlin/com/patelrmart/app/FacebookPixelPlugin.kt
package com.patelrmart.app

import android.content.Context
import android.os.Bundle
import android.util.Log
import com.facebook.FacebookSdk
import com.facebook.appevents.AppEventsLogger
import com.facebook.appevents.AppEventsConstants
import com.facebook.LoggingBehavior
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject

class FacebookPixelPlugin: FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var logger: AppEventsLogger
    private var isInitialized = false

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "facebook_pixel")
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                initializeFacebookSDK(call, result)
            }
            "logEvent" -> {
                logEvent(call, result)
            }
            "setAutoLogAppEventsEnabled" -> {
                setAutoLogAppEventsEnabled(call, result)
            }
            "setAdvertiserIDCollectionEnabled" -> {
                setAdvertiserIDCollectionEnabled(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initializeFacebookSDK(call: MethodCall, result: Result) {
        try {
            val appId = call.argument<String>("appId")
            val clientToken = call.argument<String>("clientToken")
            val pixelId = call.argument<String>("pixelId")
            val displayName = call.argument<String>("displayName")
            val enableAutoLogAppEvents = call.argument<Boolean>("enableAutoLogAppEvents") ?: true
            val enableAdvertiserIdCollection = call.argument<Boolean>("enableAdvertiserIdCollection") ?: true
            val enableCodelessEvents = call.argument<Boolean>("enableCodelessEvents") ?: true
            val enableDebugLogs = call.argument<Boolean>("enableDebugLogs") ?: false

            if (appId != null && clientToken != null) {
                // Configure Facebook SDK settings
                FacebookSdk.setApplicationId(appId)
                FacebookSdk.setClientToken(clientToken)
                FacebookSdk.setAutoLogAppEventsEnabled(enableAutoLogAppEvents)
                FacebookSdk.setAdvertiserIDCollectionEnabled(enableAdvertiserIdCollection)
                
                // Enable debug logs if requested
                if (enableDebugLogs) {
                    FacebookSdk.setIsDebugEnabled(true)
                    FacebookSdk.addLoggingBehavior(LoggingBehavior.APP_EVENTS)
                }
                
                // Initialize Facebook SDK
                FacebookSdk.sdkInitialize(context) {
                    // SDK initialized successfully
                    logger = AppEventsLogger.newLogger(context)
                    isInitialized = true
                    
                    Log.d("FacebookPixel", "✅ Facebook SDK initialized successfully")
                    Log.d("FacebookPixel", "📊 Auto Log App Events: $enableAutoLogAppEvents")
                    Log.d("FacebookPixel", "📊 Advertiser ID Collection: $enableAdvertiserIdCollection")
                    Log.d("FacebookPixel", "📊 Codeless Events: $enableCodelessEvents")
                    Log.d("FacebookPixel", "📊 Debug Logs: $enableDebugLogs")
                    
                    result.success(true)
                }
            } else {
                result.error("INVALID_PARAMETERS", "App ID and Client Token are required", null)
            }
        } catch (e: Exception) {
            Log.e("FacebookPixel", "❌ Facebook SDK initialization failed: ${e.message}")
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    private fun logEvent(call: MethodCall, result: Result) {
        if (!isInitialized) {
            result.error("NOT_INITIALIZED", "Facebook SDK not initialized", null)
            return
        }

        try {
            val eventName = call.argument<String>("eventName")
            val parameters = call.argument<Map<String, Any>>("parameters")
            val value = call.argument<Double>("value")
            val currency = call.argument<String>("currency")

            if (eventName != null) {
                // Create Bundle for parameters (following Facebook SDK best practices)
                val bundle = Bundle()
                
                // Convert parameters to Bundle
                if (parameters != null) {
                    for ((key, value) in parameters) {
                        when (value) {
                            is String -> bundle.putString(key, value)
                            is Int -> bundle.putInt(key, value)
                            is Long -> bundle.putLong(key, value)
                            is Float -> bundle.putFloat(key, value)
                            is Double -> bundle.putDouble(key, value)
                            is Boolean -> bundle.putBoolean(key, value)
                            is List<*> -> {
                                // Convert list to JSON string for content parameter
                                val jsonArray = org.json.JSONArray(value)
                                bundle.putString(key, jsonArray.toString())
                            }
                            else -> bundle.putString(key, value.toString())
                        }
                    }
                }

                // Add standard parameters if not already present
                if (value != null && !bundle.containsKey("value")) {
                    bundle.putDouble("value", value)
                }
                if (currency != null && !bundle.containsKey("currency")) {
                    bundle.putString("currency", currency)
                }

                // Log the event using Facebook SDK standard method
                logger.logEvent(eventName, value ?: 0.0, bundle)
                
                Log.d("FacebookPixel", "📊 Event logged: $eventName")
                Log.d("FacebookPixel", "📊 Value: $value")
                Log.d("FacebookPixel", "📊 Currency: $currency")
                Log.d("FacebookPixel", "📊 Parameters: $bundle")
                
                result.success(true)
            } else {
                result.error("INVALID_EVENT_NAME", "Event name is required", null)
            }
        } catch (e: Exception) {
            Log.e("FacebookPixel", "❌ Event logging failed: ${e.message}")
            result.error("LOGGING_ERROR", e.message, null)
        }
    }

    private fun setAutoLogAppEventsEnabled(call: MethodCall, result: Result) {
        try {
            val enabled = call.argument<Boolean>("enabled") ?: true
            FacebookSdk.setAutoLogAppEventsEnabled(enabled)
            Log.d("FacebookPixel", "📊 Auto Log App Events set to: $enabled")
            result.success(true)
        } catch (e: Exception) {
            Log.e("FacebookPixel", "❌ Failed to set Auto Log App Events: ${e.message}")
            result.error("SETTING_ERROR", e.message, null)
        }
    }

    private fun setAdvertiserIDCollectionEnabled(call: MethodCall, result: Result) {
        try {
            val enabled = call.argument<Boolean>("enabled") ?: true
            FacebookSdk.setAdvertiserIDCollectionEnabled(enabled)
            Log.d("FacebookPixel", "📊 Advertiser ID Collection set to: $enabled")
            result.success(true)
        } catch (e: Exception) {
            Log.e("FacebookPixel", "❌ Failed to set Advertiser ID Collection: ${e.message}")
            result.error("SETTING_ERROR", e.message, null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
} 

File: ios/Runner/FacebookPixelPlugin.swift
import Foundation
import Flutter
import FacebookCore
import FacebookLogin

@objc class FacebookPixelPlugin: NSObject, FlutterPlugin {
    private var isInitialized = false
    
    @objc static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "facebook_pixel", binaryMessenger: registrar.messenger())
        let instance = FacebookPixelPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initializeFacebookSDK(call: call, result: result)
        case "logEvent":
            logEvent(call: call, result: result)
        case "setAutoLogAppEventsEnabled":
            setAutoLogAppEventsEnabled(call: call, result: result)
        case "setAdvertiserIDCollectionEnabled":
            setAdvertiserIDCollectionEnabled(call: call, result: result)
        case "setAdvertiserTrackingEnabled":
            setAdvertiserTrackingEnabled(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initializeFacebookSDK(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let appId = args["appId"] as? String,
              let clientToken = args["clientToken"] as? String else {
            result(FlutterError(code: "INVALID_PARAMETERS", message: "App ID and Client Token are required", details: nil))
            return
        }
        
        let displayName = args["displayName"] as? String ?? "Patel's R Mart"
        let enableAutoLogAppEvents = args["enableAutoLogAppEvents"] as? Bool ?? true
        let enableAdvertiserIdCollection = args["enableAdvertiserIdCollection"] as? Bool ?? true
        let enableCodelessEvents = args["enableCodelessEvents"] as? Bool ?? true
        let enableDebugLogs = args["enableDebugLogs"] as? Bool ?? false
        
        // Configure Facebook SDK settings
        Settings.shared.appID = appId
        Settings.shared.clientToken = clientToken
        Settings.shared.displayName = displayName
        Settings.shared.isAdvertiserIDCollectionEnabled = enableAdvertiserIdCollection
        Settings.shared.isAutoLogAppEventsEnabled = enableAutoLogAppEvents
        Settings.shared.isCodelessEventsEnabled = enableCodelessEvents
        
        // Enable debug logs if requested
        if enableDebugLogs {
            Settings.shared.isDebugEnabled = true
        }
        
        // Initialize Facebook SDK
        ApplicationDelegate.shared.application(
            UIApplication.shared,
            didFinishLaunchingWithOptions: nil
        )
        
        isInitialized = true
        
        print("✅ Facebook SDK initialized successfully")
        print("📊 Auto Log App Events: \(enableAutoLogAppEvents)")
        print("📊 Advertiser ID Collection: \(enableAdvertiserIdCollection)")
        print("📊 Codeless Events: \(enableCodelessEvents)")
        print("📊 Debug Logs: \(enableDebugLogs)")
        
        result(true)
    }
    
    private func logEvent(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard isInitialized else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Facebook SDK not initialized", details: nil))
            return
        }
        
        guard let args = call.arguments as? [String: Any],
              let eventName = args["eventName"] as? String else {
            result(FlutterError(code: "INVALID_EVENT_NAME", message: "Event name is required", details: nil))
            return
        }
        
        let parameters = args["parameters"] as? [String: Any] ?? [:]
        let value = args["value"] as? Double
        let currency = args["currency"] as? String
        
        // Convert parameters to proper format using AppEvents.ParameterName
        var eventParameters: [AppEvents.ParameterName: Any] = [:]
        
        for (key, value) in parameters {
            if let stringValue = value as? String {
                eventParameters[AppEvents.ParameterName(key)] = stringValue
            } else if let numberValue = value as? NSNumber {
                eventParameters[AppEvents.ParameterName(key)] = numberValue
            } else if let boolValue = value as? Bool {
                eventParameters[AppEvents.ParameterName(key)] = boolValue
            } else if let arrayValue = value as? [Any] {
                // Convert array to JSON string for content parameter
                if let jsonData = try? JSONSerialization.data(withJSONObject: arrayValue),
                   let jsonString = String(data: jsonData, encoding: .utf8) {
                    eventParameters[AppEvents.ParameterName(key)] = jsonString
                }
            } else {
                eventParameters[AppEvents.ParameterName(key)] = "\(value)"
            }
        }
        
        // Add standard parameters if not already present
        if let value = value, !eventParameters.keys.contains(.value) {
            eventParameters[.value] = value
        }
        if let currency = currency, !eventParameters.keys.contains(.currency) {
            eventParameters[.currency] = currency
        }
        
        // Log the event using the new AppEvents.shared API
        AppEvents.shared.logEvent(AppEvents.Name(eventName), valueToSum: value, parameters: eventParameters)
        
        print("📊 Event logged: \(eventName)")
        print("📊 Value: \(value ?? 0)")
        print("📊 Currency: \(currency ?? "N/A")")
        print("📊 Parameters: \(eventParameters)")
        
        result(true)
    }
    
    private func setAutoLogAppEventsEnabled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(code: "INVALID_PARAMETERS", message: "Enabled parameter is required", details: nil))
            return
        }
        
        Settings.shared.isAutoLogAppEventsEnabled = enabled
        print("📊 Auto Log App Events set to: \(enabled)")
        result(true)
    }
    
    private func setAdvertiserIDCollectionEnabled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(code: "INVALID_PARAMETERS", message: "Enabled parameter is required", details: nil))
            return
        }
        
        Settings.shared.isAdvertiserIDCollectionEnabled = enabled
        print("📊 Advertiser ID Collection set to: \(enabled)")
        result(true)
    }
    
    private func setAdvertiserTrackingEnabled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(code: "INVALID_PARAMETERS", message: "Enabled parameter is required", details: nil))
            return
        }
        
        Settings.shared.isAdvertiserTrackingEnabled = enabled
        print("📊 Advertiser Tracking set to: \(enabled)")
        result(true)
    }
} 